### PR TITLE
add flash-attn deterministic option to flash-attn>=2.4.1

### PR DESCRIPTION
**Pull request by junrae6454**
# What does this PR do?
This PR introduces support for the deterministic mode available in Flash Attention version 2.4.1 and above. 
The deterministic mode in Flash Attention 2.4.1 ensures reproducible results, which is crucial for debugging and scientific research. By integrating this feature, users can benefit from deterministic behavior in their models.
The changes include enabling deterministic behavior through an environment variable and integrating this feature into the _flash_attention_forward function. 

* Flash Attention Forward Function Update:
  * Added deterministic parameter to _flash_attention_forward to control deterministic behavior.
  * Modified the function to check for Flash Attention version 2.4.1 or higher and set the deterministic argument accordingly.

* Trainer Utilities Update:
  * Set the FLASH_ATTENTION_DETERMINISTIC environment variable in enable_full_determinism.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@ArthurZucker

### CODE REVIEW

File: src/transformers/modeling_flash_attention_utils.py [hunk 1]
```diff
@@ -14,12 +14,13 @@
...
 
 import inspect
+import os
 from typing import Optional, Tuple
 
 import torch
 import torch.nn.functional as F
 
-from .utils import is_flash_attn_2_available
+from .utils import is_flash_attn_2_available, is_flash_attn_greater_or_equal_2_41
```
Comment (ArthurZucker) [inline src/transformers/modeling_flash_attention_utils.py:23]:
```suggestion
from .utils import is_flash_attn_2_available, is_flash_attn_greater
```

File: src/transformers/modeling_flash_attention_utils.py [hunk 2]
```diff
@@ -177,6 +181,9 @@ def _flash_attention_forward(
     )
     flash_kwargs = {"window_size": (sliding_window, sliding_window)} if use_sliding_windows else {}
 
+    if is_flash_attn_greater_or_equal_2_41():
```
Comment (ArthurZucker) [inline src/transformers/modeling_flash_attention_utils.py:184]:
```suggestion
    if is_flash_attn_greater_or_equal("2.41"):
```

File: src/transformers/utils/import_utils.py [hunk 3]
```diff
@@ -814,6 +814,13 @@ def is_flash_attn_greater_or_equal_2_10():
     return version.parse(importlib.metadata.version("flash_attn")) >= version.parse("2.1.0")
 
 
+def is_flash_attn_greater_or_equal_2_41():
+    if not _is_package_available("flash_attn"):
+        return False
+
+    return version.parse(importlib.metadata.version("flash_attn")) >= version.parse("2.4.1")
```
Comment (ArthurZucker) [inline src/transformers/utils/import_utils.py:821]:
```suggestion
```
unnecessary, see `def is_flash_attn_greater_or_equal(library_version: str):`

**Review by ArthurZucker — COMMENTED** [review]
sounds good! are there other args we should be reading from env ?

**Comment by junrae6454** [comment]
> sounds good! are there other args we should be reading from env ?

@ArthurZucker 
Other changes are related to the model structure and the model's return value, so there doesn't seem to be anything else that needs to be read from the environment.

**Review by ghost — APPROVED** [review]

**Review by ArthurZucker — APPROVED** [review]
thanks for iterating

**Pull request merged**